content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
+
+Contrary to the name "optimizations," these algorithms don't seek to optimize, or find the best possible program for the metric. Given this misnomer, Alexia Massalin coined the term superoptimizers.
+
+Given an instruction set, the superoptimizer finds the shortest program to compute a function. [3]
+
+Most problems within the field of superoptimizations are intractable, and therefore we generally employ exhaustive bottom-up search or stochastic search.
+
+In this paper, we apply stochastic methods to find superoptimizations of loop-free programs. Similar to STOKE [8], we use MarkovChain Monte-Carlo (MCMC) methods to conduct a random walk on the space of programs.
+
+In contrast with STOKE, which uses a twostage process to introduce the program performance component, our BLOKE implementation gradually introduces the performance component into the MCMC search.
+We will first give a primer on Bril, the Metropolis Hastings algorithm, and STOKE, the precursor to BLOKE. We then describe BLOKE, our stochastic superoptimizer for loop-free Bril programs. Finally, we evaluate BLOKE to show that BLOKE can find superoptimizations of Bril programs scalably.
+
+## Bril
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-01.jpg?height=673&width=569&top_left_y=1342&top_left_x=1189)
+
+Figure 1: clobber Benchmark
+
+For ease of use, Bril uses a JSON format and has Typescript and Rust commandline interpreters that support a profiling flag "-p" to count the number of dynamic instructions executed at runtime.
+
+## MCMC Sampling
+
+Rejection sampling methods are in general intractable on high-dimensional probabilistic models where the sample space increases exponentially with the number of dimensions. For the superoptimization problem, even if we limit the length of the program, the number of possible programs is exponential with respect to the number of possible instructions and operands.
+
+Furthermore, MC sampling requires us to draw independent samples. This is counterintuitive when as an input to a superoptimization program, we are given a program to optimize. We could instead use the original program as a prior to make our acceptance decisions.
+
+MCMC addresses these issues through the use of a Markov Chain, where a sample is dependent on the prior samples. Markov Chains solve the curse of dimensionality through limiting the sample space at each iteration to a mutation of the previous samples.
+
+## Metropolis-Hastings Algorithm
+
+Markov Chain Monte Carlo (MCMC) methods draw samples from previous samples. The Metropolis-Hastings algorithm is a MCMC method, where for every MCMC iteration, we base our sample based on the previous sample.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-02.jpg?height=717&width=686&top_left_y=1681&top_left_x=235)
```

Comment:
This image is again just text—it would be more readable if it were included as actual text here. If you really want the LaTeX `algorithm`-style formatting, can you please include the image here in the repo instead of hosted externally?